deepspeed/pt/deepspeed_zero_optimizer.py

```diff
@@ -95,10 +96,45 @@ def move_to_cpu(tensor_list):
         tensor.data = tensor.data.cpu()
 
 
+def move_to_cuda(tensor_list, cuda_device):
+    for tensor in tensor_list:
+        tensor.data = tensor.data.to(cuda_device)
+
+
 def print_rank_msg(msg):
     print(f"rank {dist.get_rank()} - {msg}")
 
 
+def async_migrate_to(obj, dev, main_stream=None):
+    if torch.is_tensor(obj):
+        obj = Variable(obj)
+    if isinstance(obj, Variable):
+        v = obj.cuda(dev, async=True)
+        if main_stream is not None:
+            v.data.record_stream(main_stream)
+        return v
+    elif isinstance(obj, collections.Mapping):
+        return {k: async_copy_to(o, dev, main_stream) for k, o in obj.items()}
+    elif isinstance(obj, collections.Sequence):
+        return [async_copy_to(o, dev, main_stream) for o in obj]
+    else:
+        return obj
+
+
+def async_copy_to(obj, dev, main_stream=None):
+    if torch.is_tensor(obj):
+        obj = Variable(obj)
+    if isinstance(obj, Variable):
+        target = torch.empty_like(obj, device=dev).copy_(obj)
+        if main_stream is not None:
+            target.data.record_stream(main_stream)

```

Comment:
This is just a curiosity question: Do you need record_stream if you are doing explicity wait on this stream before accessing or deleting the target tensor?

Reply:
Thank you for pointing this out! I was planning to make line 128 as
target = torch.empty_like(obj, device=dev).copy_(obj, non_blocking=True)
for async data copy

Reply:
@samyam, I think we need the record_stream here. According to the [documentation](https://pytorch.org/docs/1.5.1/tensors.html?highlight=record_stream#torch.Tensor.record_stream), record_stream is the safe way to pass tensors from an allocating stream to a consuming stream. Jie and I agree that the `self.cpu_computation_stream` should passed in as the consuming stream.
On the other hand, the [wait_stream](https://pytorch.org/docs/1.5.1/cuda.html?highlight=wait_stream#torch.cuda.Stream.wait_stream) is needed to synchronize two streams in general.